src/gateway/shard_manager.rs

```diff
@@ -0,0 +1,303 @@
+use futures::{future, Future, Stream, Poll, Sink, StartSend, AsyncSink};
+use ::Error;
+use std::collections::{VecDeque, HashMap};
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::time::Duration;
+use gateway::{
+    shard::Shard,
+    queue::ReconnectQueue,
+};
+use model::event::{Event, GatewayEvent};
+use tokio_core::reactor::Handle;
+use tokio_timer::Timer;
+use futures::sync::mpsc::{
+    unbounded, UnboundedSender, UnboundedReceiver, 
+    channel, Sender as MpscSender, Receiver as MpscReceiver,
+    SendError,
+};
+use tungstenite::{Message as TungsteniteMessage, Error as TungsteniteError};
+
+#[derive(Clone, Copy, Debug)]
+pub enum ShardingStrategy {
+    Autoshard,
+    Range(u64, u64, u64),
+}
+
+impl ShardingStrategy {
+    pub fn auto() -> Self {
+        ShardingStrategy::Autoshard
+    }
+
+    pub fn multi(count: u64) -> Self {
+        ShardingStrategy::Range(0, count, count)
+    }
+
+    pub fn simple() -> Self {
+        ShardingStrategy::Range(0, 1, 1)
+    }
+
+    pub fn range(index: u64, count: u64, total: u64) -> Self {
+        ShardingStrategy::Range(index, count, total)
+    }
+}
+
+impl Default for ShardingStrategy {
+    fn default() -> Self {
+        ShardingStrategy::Autoshard
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct ShardManagerOptions<T: ReconnectQueue> {
+    pub strategy: ShardingStrategy,
+    pub token: Rc<String>,
+    pub ws_uri: Rc<String>,
+    pub queue: T,
+}
+
+pub type WrappedShard = Rc<RefCell<Shard>>;
+pub type Message = (WrappedShard, TungsteniteMessage);
+pub type MessageStream = UnboundedReceiver<Message>;
+type ShardsMap = Rc<RefCell<HashMap<u64, WrappedShard>>>;
+
+pub struct ShardManager<T: ReconnectQueue> {
+    pub queue: VecDeque<u64>,
+    reconnect_queue: T,
+    shards: ShardsMap,
+    pub strategy: ShardingStrategy,
+    pub token: Rc<String>,
+    pub ws_uri: Rc<String>,
+    handle: Handle,
+    message_stream: Option<MessageStream>,
+    queue_sender: MpscSender<u64>,
+    queue_receiver: Option<MpscReceiver<u64>>,
+    #[allow(dead_code)]
+    non_exhaustive: (),
```

Comment:
We don't actually need this non-exhaustive field if we have other private fields.